### PR TITLE
baseapp-profiles: edit member role

### PR DESCRIPTION
**Pull request by hpmoreira05**
Description

As a user, on the BaseApp Profile,I would like to Change the role of a member, In order to to give them permissions for different managing actions inside the profile.

 

Original Story:  
Acceptance Criteria 
Context
In this story we want to enable to set permissions to members in an organization. 

Business Rules - Edit Member Role - 3

Given a user wants to change the role or remove a member from their  organization, they will be able to click on a dropdown next to the members name in the Members List

The user is able to change between Admin and Manager Role

Given the user has selected a new role for that member, display a confirmation dialog to make sure he intends to do this change.

Only show this confirmation dialog when changing to Admin

Given the user has confirmed the change of role, then the member's role should now reflect it in the FE and BE.

Only admin roles are able to interact with the Member's editing dropdown

Managers will have this dropdown disable.

Add the permission the following permissions to the Admin role:

Editing the role of a member

Removing a member

Copy Text

Role Change Confirmation: Are you sure you want to assign the [Role] role to this user?

Removal Confirmation:  Are you sure you want to remove this member? This action will revoke access to the organizational profile 

Current behavior 

Expected behavior 

Code Snippet 

Design Link: https://www.figma.com/design/XRD6wSl1m8Kz6XUcAy5CLp/BaseApp---WEB?node-id=3790-91705&node-type=instance&t=4tZzXNioOJs6iIeL-0 

Template: https://bitbucket.org/silverlogic/baseapp-frontend-template/pull-requests/122
BA FE Package: https://github.com/silverlogic/baseapp-frontend/pull/143

## Summary by CodeRabbit

- **New Features**
	- Introduced a new mutation for updating user roles within profiles.
	- Added new permission checks related to profile management.

- **Bug Fixes**
	- Enhanced error handling for role updates and permission checks.

- **Tests**
	- Implemented tests for validating role update functionality, including permission scenarios.

- **Chores**
	- Updated the version number of the `baseapp_profiles` package from 0.3.5 to 0.3.6.

### CODE REVIEW

File: baseapp-profiles/baseapp_profiles/graphql/mutations.py [hunk 1]
```diff
@@ -114,6 +114,43 @@ def perform_mutate(cls, serializer, info):
         )
 
 
+RoleTypeEnum = graphene.Enum.from_enum(ProfileUserRole.ProfileRoles)
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/graphql/mutations.py:117]: instead of defining it here again do `from .object_types import ProfileRoleTypesEnum`

**Review by nossila — CHANGES_REQUESTED** [review]

File: baseapp-profiles/baseapp_profiles/graphql/object_types.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ def resolve_metadata(cls, instance, info):
         return ProfileMetadata(instance, info)
 
     @classmethod
-    def resolve_members(cls, instance, info, **kwargs):
+    def resolve_members(cls, instance, info, order_by_status=None, **kwargs):
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/graphql/object_types.py:134]: `order_by_status` is not needed here

File: baseapp-profiles/baseapp_profiles/tests/test_graphql_mutations_update.py [hunk 3]
```diff
@@ -32,6 +34,50 @@
     }
 """
 
+# PROFILE_ROLE_UPDATE_GRAPHQL = """
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/tests/test_graphql_mutations_update.py:37]: remove commented code

File: baseapp-profiles/baseapp_profiles/graphql/mutations.py [hunk 4]
```diff
@@ -112,7 +112,42 @@
...
+        profile_id = input.get("profile_id")
+        role_type = input.get("role_type")
+        user_pk = get_pk_from_relay_id(user_id)
+        profile_pk = get_pk_from_relay_id(profile_id)
+        
+
+        try:
+            obj = ProfileUserRole.objects.get(user_id=user_pk, profile_id=profile_pk)
+        except ProfileUserRole.DoesNotExist:
+            raise ValueError(_("Role not found"))
```
[the remark this answers is not in the document]
Comment (nick-prat) [reply baseapp-profiles/baseapp_profiles/graphql/mutations.py:139]: I didn't know about this python feature :thinking:, but it's actually pretty nice, not sure why it's recommending `raise .. from None` though, it should be
```python
except ProfileUserRole.DoesNotExist as e:
    raise ValueError(_("Role not found")) from e
```

File: baseapp-profiles/baseapp_profiles/permissions.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import swapper
 from django.contrib.auth.backends import BaseBackend
 
+from .models import ProfileUserRole
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/permissions.py:4]: you need to import using swapper:
```python
ProfileUserRole = swapper.load_model("baseapp_profiles", "ProfileUserRole")
```

File: baseapp-profiles/baseapp_profiles/graphql/mutations.py [hunk 6]
```diff
@@ -210,3 +246,4 @@ class ProfilesMutations(object):
     profile_create = ProfileCreate.Field()
     profile_update = ProfileUpdate.Field()
     profile_delete = ProfileDelete.Field()
+    role_update = RoleUpdate.Field()
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/graphql/mutations.py:249]: change to `profile_role_update` to make it more organized and explicit about what this is about in the graphql mutations

File: baseapp-profiles/baseapp_profiles/tests/test_graphql_mutations_update.py [hunk 7]
```diff
@@ -199,3 +217,82 @@ def test_user_with_permission_can_update_profile(django_user_client, graphql_use
...
+    ProfileUserRoleFactory(profile=profile, user=user_3, role=ProfileUserRole.ProfileRoles.MANAGER)
+
+    response = graphql_user_client(
+        PROFILE_ROLE_UPDATE_GRAPHQL,
+        variables={
+            "input": {"userId": user_3.relay_id, "profileId": profile.relay_id, "roleType": "ADMIN"}
+        },
+    )
+    content = response.json()
+    print(content)
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/tests/test_graphql_mutations_update.py:296]: remove print please

File: baseapp-profiles/baseapp_profiles/graphql/mutations.py [hunk 8]
```diff
@@ -120,6 +122,40 @@ def perform_mutate(cls, serializer, info):
...
+        user_id = input.get("user_id")
+        profile_id = input.get("profile_id")
+        role_type = input.get("role_type")
+        user_pk = get_pk_from_relay_id(user_id)
+        profile_pk = get_pk_from_relay_id(profile_id)
+
+        try:
+            obj = ProfileUserRole.objects.get(user_id=user_pk, profile_id=profile_pk)
+        except ProfileUserRole.DoesNotExist:
+            raise ValueError(_("Role not found"))
```
Comment (nossila) [inline baseapp-profiles/baseapp_profiles/graphql/mutations.py:145]: raise a `GraphQLError` instead to make it easier to handle in the frontend if necessary

**Review by nossila — CHANGES_REQUESTED** [review]

**Review by nossila — APPROVED** [review]

**Review by nick-prat — APPROVED** [review]

**Review by vitorguima — APPROVED** [review]

**Pull request merged**